### ISSUE TITLE
[JENKINS-60434] "Prepare for shutdown" should continue executing already running pipelines to completion

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/workflow-cps-plugin#1553]
<p>Based on <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum" class="user-hover" rel="dnusbaum">dnusbaum</a>'s comment from <a href="https://issue-redirect.jenkins.io/issue/34256" title="Preparing Jenkins For Shutdown Hangs Running Pipelines" class="issue-link" data-issue-key="JENKINS-34256"><del>JENKINS-34256</del></a>:</p>
<blockquote>
<p>A fix for this issue was just released in Pipeline: Groovy Plugin version 2.78. I think there is/was some confusion as to the expected behavior (myself included!), so let me try to clarify: When Jenkins prepares for shutdown, all running Pipelines are paused, and this is the intended behavior. The unintended behavior was that if you canceled shutdown, Pipelines remained paused. This has been fixed in 2.78; Pipelines will now resume execution if shutdown is canceled. Before 2.78, you had to manually pause and unpause each Pipeline to get it to resume execution, or restart Jenkins. Additionally, preparing Jenkins for shutdown and canceling shutdown now each cause a message to be printed to Pipeline build logs indicating that the Pipeline is being paused or resumed due to shutdown so that it is easier to understand what is happening.</p>

<p>Based on comments here and elsewhere, I think some users would prefer a variant of "Prepare for shutdown" in which Pipelines continue executing to completion, the same as other types of jobs like Freestyle. If that is something you want, please open a new ticket, describing your use case and the desired behavior.</p>

<p><span class="error">&#91;...&#93;</span></p>

<p>If there is some other aspect of this issue that you would like to see addressed, or a different behavior you would prefer, please open a new ticket describing your particular use case.</p></blockquote>

<p>My use case is to <b>make restarting Jenkins master to allow upgrading Jenkins core or updating Jenkins plugins easier</b>, because now I need to do the following:</p>
<ol>
	<li>wait until no pipelines are running anymore
	<ul>
		<li>which can get difficult in bigger Jenkins environments during the day i.e. normal working hours (due to steady commits triggering pipelines), but also in case there are longer lasting test suites that e.g. are triggered all around the clock</li>
	</ul>
	</li>
	<li>click "prepare for shutdown"</li>
	<li>... (continue normal work like upgrading/updating)</li>
</ol>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=reinholdfuereder">reinholdfuereder</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-60434" target="_blank">"Prepare for shutdown" should continue executing already running pipelines to completion</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: workflow-cps-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 42
<li><b>watchers</b>: 50
<li><b>imported</b>: 2025-12-07
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Based on <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum" class="user-hover" rel="dnusbaum">dnusbaum</a>'s comment from <a href="https://issue-redirect.jenkins.io/issue/34256" title="Preparing Jenkins For Shutdown Hangs Running Pipelines" class="issue-link" data-issue-key="JENKINS-34256"><del>JENKINS-34256</del></a>:</p>
<blockquote>
<p>A fix for this issue was just released in Pipeline: Groovy Plugin version 2.78. I think there is/was some confusion as to the expected behavior (myself included!), so let me try to clarify: When Jenkins prepares for shutdown, all running Pipelines are paused, and this is the intended behavior. The unintended behavior was that if you canceled shutdown, Pipelines remained paused. This has been fixed in 2.78; Pipelines will now resume execution if shutdown is canceled. Before 2.78, you had to manually pause and unpause each Pipeline to get it to resume execution, or restart Jenkins. Additionally, preparing Jenkins for shutdown and canceling shutdown now each cause a message to be printed to Pipeline build logs indicating that the Pipeline is being paused or resumed due to shutdown so that it is easier to understand what is happening.</p>

<p>Based on comments here and elsewhere, I think some users would prefer a variant of "Prepare for shutdown" in which Pipelines continue executing to completion, the same as other types of jobs like Freestyle. If that is something you want, please open a new ticket, describing your use case and the desired behavior.</p>

<p><span class="error">&#91;...&#93;</span></p>

<p>If there is some other aspect of this issue that you would like to see addressed, or a different behavior you would prefer, please open a new ticket describing your particular use case.</p></blockquote>

<p>My use case is to <b>make restarting Jenkins master to allow upgrading Jenkins core or updating Jenkins plugins easier</b>, because now I need to do the following:</p>
<ol>
	<li>wait until no pipelines are running anymore
	<ul>
		<li>which can get difficult in bigger Jenkins environments during the day i.e. normal working hours (due to steady commits triggering pipelines), but also in case there are longer lasting test suites that e.g. are triggered all around the clock</li>
	</ul>
	</li>
	<li>click "prepare for shutdown"</li>
	<li>... (continue normal work like upgrading/updating)</li>
</ol>
</pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>vtwaldo21:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-60434?focusedId=409211&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-409211">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>@Tim Brown, I think this whole ticket is based on the fact that &lt;jenkins_url&gt;/safeRestart doesn't actually work.  I believe it actually does work in some cases but not in others.</p>

<p>For example, I hit this bug at least once a week when applying patches. I think our cause may be that we have some freestyle jobs that trigger pipeline jobs (the horror!). Since the pipeline jobs pause indefinitely, the freestyle jobs never complete and the restart will never happen.</p>

<p>I think the more common use case would be /safeRestart should just let all running jobs finish as it has done since the beginning. The only case to pause running jobs would be if you had one of those 18 hour jobs running and had to restart immediately.  </p>

<p>I agree completely with all 4 of @Jason Antman's points as well.</p>
</pre>
</details>

</ul>
</details>

<p>@​Tim Brown, I think this whole ticket is based on the fact that <jenkins_url>/safeRestart doesn't actually work.  I believe it actually does work in some cases but not in others.</p>

<p>For example, I hit this bug at least once a week when applying patches. I think our cause may be that we have some freestyle jobs that trigger pipeline jobs (the horror!). Since the pipeline jobs pause indefinitely, the freestyle jobs never complete and the restart will never happen.</p>

<p>I think the more common use case would be /safeRestart should just let all running jobs finish as it has done since the beginning. The only case to pause running jobs would be if you had one of those 18 hour jobs running and had to restart immediately.  </p>

<p>I agree completely with all 4 of @​Jason Antman's points as well.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>paybas:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-60434?focusedId=428905&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-428905">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yeah this has been plaguing me forever. The "Prepare for Shutdown" feature not allowing already running pipelines to continue/finish is a major p.i.t.a.</p>

<p>I need a way to put Jenkins in maintenance mode at 17:00 to prevent new builds from being started, but let current ones finish so I have a clean Jenkins at 18:00 when I actually start the maintenance. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/sad.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>Yeah this has been plaguing me forever. The "Prepare for Shutdown" feature not allowing already running pipelines to continue/finish is a major p.i.t.a.</p>

<p>I need a way to put Jenkins in maintenance mode at 17:00 to prevent new builds from being started, but let current ones finish so I have a clean Jenkins at 18:00 when I actually start the maintenance. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/sad.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ivanrossier:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-60434?focusedId=428919&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-428919">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Totally agree with last remark: "Prepare for shutdown' was really useful for maintenance  purpose.</p>
</pre>
</details>

</ul>
</details>

<p>Totally agree with last remark: "Prepare for shutdown' was really useful for maintenance  purpose.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kastningj.png" title="kastningj's avatar" /> kastningj:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-60434?focusedId=428962&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-428962">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yes, 100% agree! Please make fixing the "Prepare for shutdown" a priority!</p>
</pre>
</details>

</ul>
</details>

<p>Yes, 100% agree! Please make fixing the "Prepare for shutdown" a priority!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>karniemi:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-60434?focusedId=435245&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-435245">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The official KB from Cloudbees also states it incorrectly: <a href="https://docs.cloudbees.com/docs/cloudbees-ci-kb/latest/client-and-managed-masters/how-do-i-stop-builds-on-slaves-to-prepare-for-routine-jenkins-maintenance" class="external-link" target="_blank" rel="nofollow noopener">https://docs.cloudbees.com/docs/cloudbees-ci-kb/latest/client-and-managed-masters/how-do-i-stop-builds-on-slaves-to-prepare-for-routine-jenkins-maintenance</a></p>
</pre>
</details>

</ul>
</details>

<p>The official KB from Cloudbees also states it incorrectly: <a href="https://docs.cloudbees.com/docs/cloudbees-ci-kb/latest/client-and-managed-masters/how-do-i-stop-builds-on-slaves-to-prepare-for-routine-jenkins-maintenance" class="external-link" target="_blank" rel="nofollow noopener">https://docs.cloudbees.com/docs/cloudbees-ci-kb/latest/client-and-managed-masters/how-do-i-stop-builds-on-slaves-to-prepare-for-routine-jenkins-maintenance</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>karniemi:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-60434?focusedId=441780&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-441780">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><del>Another possible work around described here: <a href="https://issue-redirect.jenkins.io/issue/72097" title="&quot;Run Exclusive&quot; does not work after Jenkins restart/prepareForShutdown" class="issue-link" data-issue-key="JENKINS-72097">JENKINS-72097</a> "Run Exclusive" does not work after Jenkins restart/prepareForShutdown - Jenkins Jira</del></p>

<p><del>I've been looking for pretty much all the alternatives to tackle this proble ... but all have some short-comings. That is now the closest I've found and does not require adding any new logic nor plugins to all the jenkins jobs.</del></p>

<p> </p>

<p>Edit: It's BS. That solution does not work either. Despite the docs of the plug-in, the Jenkins-pipelines get paused at next node()-section if a "Run Exclusive"-job is running. I'm considering to go for abruptly aborting all jobs and rebooting all jenkins-nodes when the maintenance breaks start - nevermind the devs and the running builds.</p>
</pre>
</details>

</ul>
</details>

<p><del>Another possible work around described here: <a href="https://issue-redirect.jenkins.io/issue/72097" title=""Run Exclusive" does not work after Jenkins restart/prepareForShutdown" class="issue-link" data-issue-key="JENKINS-72097">JENKINS-72097</a> "Run Exclusive" does not work after Jenkins restart/prepareForShutdown - Jenkins Jira</del></p>

<p><del>I've been looking for pretty much all the alternatives to tackle this proble ... but all have some short-comings. That is now the closest I've found and does not require adding any new logic nor plugins to all the jenkins jobs.</del></p>

<p> </p>

<p>Edit: It's BS. That solution does not work either. Despite the docs of the plug-in, the Jenkins-pipelines get paused at next node()-section if a "Run Exclusive"-job is running. I'm considering to go for abruptly aborting all jobs and rebooting all jenkins-nodes when the maintenance breaks start - nevermind the devs and the running builds.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ajkyffin:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-60434?focusedId=445076&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-445076">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I've found this frustrating because jobs often don't survive a restart.</p>

<p>I have a solution which allows you to disable the pausing of pipelines by setting a system property: <a href="https://github.com/jenkinsci/workflow-cps-plugin/pull/846" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/workflow-cps-plugin/pull/846</a>.</p>
</pre>
</details>

</ul>
</details>

<p>I've found this frustrating because jobs often don't survive a restart.</p>

<p>I have a solution which allows you to disable the pausing of pipelines by setting a system property: <a href="https://github.com/jenkinsci/workflow-cps-plugin/pull/846" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/workflow-cps-plugin/pull/846</a>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-60434?focusedId=445077&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-445077">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>The only case to pause running jobs would be if you had one of those 18 hour jobs running and had to restart immediately.</p></blockquote>

<p>Not really. The behavior is designed to allow the controller to be restarted nearly immediately, even when there are Pipeline builds running. Any builds running inside a <tt>sh</tt> step (for example) may be “paused” at the Groovy level but the actual task running on the agent continues without interruption and may complete during the quiet period, while the controller is restarting, or after restart, without affecting ultimate build status. Once all builds get to a safe spot (which should normally be in a matter of seconds, assuming there are not any freestyle builds running) the restart can proceed.</p>

<p>There probably needs to be a distinct admin gesture to “prepare for <em>eventual</em> shutdown” to handle special circumstances, such as:</p>
<ul>
	<li>there are freestyle (or other non-Pipeline) builds running, which cannot tolerate a controller restart</li>
	<li>there some Pipeline builds running which are marked with the option to not permit resumption across controller restarts</li>
</ul>


<p>This would need to suppress the behavior of pausing the CPS VM portion of running Pipeline builds <em>and</em> force <tt>CpsFlowExecution.blocksRestart</tt> on so that the restart would wait until all builds of all types have completed naturally. As I recall there is also logic to suppress scheduling of new queue items, which would need to exempt new <tt>node</tt> blocks from running Pipeline builds (or else the system would livelock).</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>The only case to pause running jobs would be if you had one of those 18 hour jobs running and had to restart immediately.</p></blockquote>

<p>Not really. The behavior is designed to allow the controller to be restarted nearly immediately, even when there are Pipeline builds running. Any builds running inside a <tt>sh</tt> step (for example) may be “paused” at the Groovy level but the actual task running on the agent continues without interruption and may complete during the quiet period, while the controller is restarting, or after restart, without affecting ultimate build status. Once all builds get to a safe spot (which should normally be in a matter of seconds, assuming there are not any freestyle builds running) the restart can proceed.</p>

<p>There probably needs to be a distinct admin gesture to “prepare for <em>eventual</em> shutdown” to handle special circumstances, such as:</p>
<ul>
	<li>there are freestyle (or other non-Pipeline) builds running, which cannot tolerate a controller restart</li>
	<li>there some Pipeline builds running which are marked with the option to not permit resumption across controller restarts</li>
</ul>


<p>This would need to suppress the behavior of pausing the CPS VM portion of running Pipeline builds <em>and</em> force <tt>CpsFlowExecution.blocksRestart</tt> on so that the restart would wait until all builds of all types have completed naturally. As I recall there is also logic to suppress scheduling of new queue items, which would need to exempt new <tt>node</tt> blocks from running Pipeline builds (or else the system would livelock).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>romanz:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-60434?focusedId=445106&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-445106">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This issue is quite complicated in our setup because we have many cascaded jobs and lots of them don't survive a restart (for one or another reason).
So I could think of 2 possibilities to solve this:</p>
<ul>
	<li>inhibit "external" triggers: don't allow things like manually (re)starting a build, SCM trigger, time trigger,... but still allow subjobs to be executed.
This would allow cascaded jobs to finish (as they need to execute subjobs to get finished).
OR</li>
	<li>don't start any new jobs AND wait until all <em>running</em> jobs are in a state where they are waiting for a subjob to be finished - presuming that this is a good state for a safe restart in any case.</li>
</ul>


<p>I don't know if any of this would be easy (or even possible) to implement.
And of course it still leaves the inconvenience that you have to wait until long running jobs get finished (if any) but it would help in our case.</p>
</pre>
</details>

</ul>
</details>

<p>This issue is quite complicated in our setup because we have many cascaded jobs and lots of them don't survive a restart (for one or another reason).<br/>
So I could think of 2 possibilities to solve this:</p>
<ul>
	<li>inhibit "external" triggers: don't allow things like manually (re)starting a build, SCM trigger, time trigger,... but still allow subjobs to be executed.<br/>
This would allow cascaded jobs to finish (as they need to execute subjobs to get finished).<br/>
OR</li>
	<li>don't start any new jobs AND wait until all <em>running</em> jobs are in a state where they are waiting for a subjob to be finished - presuming that this is a good state for a safe restart in any case.</li>
</ul>


<p>I don't know if any of this would be easy (or even possible) to implement.<br/>
And of course it still leaves the inconvenience that you have to wait until long running jobs get finished (if any) but it would help in our case.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ajkyffin:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-60434?focusedId=445175&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-445175">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>To allow pipelines to run to completion, including new node steps, <tt>ContinuedTask</tt> would have to extend <tt>Task.NonBlockingTask</tt> to allow them to be scheduled. <tt>CpsFlowExecution.blocksRestart()</tt> could simply return <tt>true</tt>. This failed for me with cloud executors because no new agents were provisioned during quietDown.</p>

<p>Allowing pipelines to run until the next node step requires not pausing the pipeline. <tt>CpsFlowExecution.blocksRestart()</tt> already checks with each <tt>StepExecution</tt> so <tt>ExecutorStepExecution.blocksRestart()</tt> could return <tt>true</tt> unless it itself is blocked. However, the pipeline then has to be paused to save its state before Jenkins can be restarted.</p>

<p>I think both approaches would fail in the case of nested jobs.</p>
</pre>
</details>

</ul>
</details>

<p>To allow pipelines to run to completion, including new node steps, <tt>ContinuedTask</tt> would have to extend <tt>Task.NonBlockingTask</tt> to allow them to be scheduled. <tt>CpsFlowExecution.blocksRestart()</tt> could simply return <tt>true</tt>. This failed for me with cloud executors because no new agents were provisioned during quietDown.</p>

<p>Allowing pipelines to run until the next node step requires not pausing the pipeline. <tt>CpsFlowExecution.blocksRestart()</tt> already checks with each <tt>StepExecution</tt> so <tt>ExecutorStepExecution.blocksRestart()</tt> could return <tt>true</tt> unless it itself is blocked. However, the pipeline then has to be paused to save its state before Jenkins can be restarted.</p>

<p>I think both approaches would fail in the case of nested jobs.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-60434?focusedId=445224&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-445224">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>a distinct admin gesture</p></blockquote>

<p>Unnecessary I guess, if there are any running non-Pipeline builds or Pipeline builds marked with <b>Do not allow the pipeline to resume if the controller restarts</b>: this should be a sufficient signal.</p>

<p>As mentioned in recent comments, there would be some work to do to ensure that new <tt>node</tt> blocks could be scheduled but not new top-level builds…except perhaps new downstream builds triggered via the <tt>build</tt> step (with the default <tt>wait: true</tt>), since otherwise you would again livelock.</p>

<p>The trickier question is the case that there is a <em>mixture</em> of resumable and non-resumable builds. The current behavior optimizes for a quicker restart, by pausing new activity in the resumable Pipeline builds. But if the non-resumable builds, currently freestyle, would still be running for a long time anyway then you may as well get more work done in the resumable builds while you wait. You just do not want to be initiating new agent connections and the like right before the controller is about to shut down. Perhaps it would make sense to wait for a few minutes to see if the safe restart proceeds in a timely fashion, before giving up and unpausing any resumable builds.</p>

<p>Of course it would also be valuable to track down cases of Pipeline builds which ought to survive restarts (i.e., do not involve weird Groovy logic with non-<tt>Serializable</tt> local variables!) but sometimes do not, come up with reproducible test cases, and get those fixed.</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>a distinct admin gesture</p></blockquote>

<p>Unnecessary I guess, if there are any running non-Pipeline builds or Pipeline builds marked with <b>Do not allow the pipeline to resume if the controller restarts</b>: this should be a sufficient signal.</p>

<p>As mentioned in recent comments, there would be some work to do to ensure that new <tt>node</tt> blocks could be scheduled but not new top-level builds…except perhaps new downstream builds triggered via the <tt>build</tt> step (with the default <tt>wait: true</tt>), since otherwise you would again livelock.</p>

<p>The trickier question is the case that there is a <em>mixture</em> of resumable and non-resumable builds. The current behavior optimizes for a quicker restart, by pausing new activity in the resumable Pipeline builds. But if the non-resumable builds, currently freestyle, would still be running for a long time anyway then you may as well get more work done in the resumable builds while you wait. You just do not want to be initiating new agent connections and the like right before the controller is about to shut down. Perhaps it would make sense to wait for a few minutes to see if the safe restart proceeds in a timely fashion, before giving up and unpausing any resumable builds.</p>

<p>Of course it would also be valuable to track down cases of Pipeline builds which ought to survive restarts (i.e., do not involve weird Groovy logic with non-<tt>Serializable</tt> local variables!) but sometimes do not, come up with reproducible test cases, and get those fixed.</p>

**Comment by SampsonCrowley** [comment]
this was opened 4+ years ago, any updates or news?
